### PR TITLE
INT- B 21608 alaska update search counseling office query

### PR DESCRIPTION
**Pull request by MInthavongsay**
## Summary

This PR addresses the search counseling office issue when AK duty locations were migrated into the system from [B-21481](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1016235) where counseling office dropdown was not being populated when an AK duty location is selected.

NOTE:  This PR does not address NSRA15 associated duty locations that were migrated. Side effect is that if a known NSRA15 location is selected, the search counseling offices FAILS and the dropdown does not get populated. A [BL](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1031218) for this particular issue has been created to resolve this issue. 

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. From the customer app, create users with following affiliations: AIR FORCE, SPACE FORCE, ARMY, MARINES, NAVY, COAST GUARDS.
2. For each user login and start the move onboarding process.  Search for known AK location names using file: 
[GOOD-AK-data-1731517877769.csv](https://github.com/user-attachments/files/17738686/GOOD-AK-data-1731517877769.csv) . Click selection and verify Counseling Office dropdown gets populated. NOTE: there is an existing UI bug where if there is a search error, the dropdown will not clear the dropdown but will retain the dropdown state from the last successful search. This means that even if it fails, you might see a FALSE positive.  Verify logs for every search! A duty location's counseling office is retrieved using its associative GBLOC for that address. Unlike the CONUS duty locations, AK locations will have another layer of business logic to determine what GLOC to use based on the ZONE the address location is assigned to. There are 5 known Zones -->  Zone 1 to  5.   Here are the business rules per Zone:

- **Zone I** shipments will be processed by JPPSO Elmendorf-Richardson (GBLOC MBFL).
- **Zone II** shipments are divided between JPPSO-NW (GBLOC JEAT) and JPPSO Elmendorf-Richardson (GBLOC MBFL) in the following manner:  JPPSO-NW (GBLOC JEAT) will process all shipments that are aligned with Zone II excluding Air Force and Space Force personnel. Air Force and Space Force personnel in Zone II will be processed by JPPSO Elmendorf-Richardson (GBLOC MBFL).
- **Zone III** shipments will be processed by USCG Base Kodiak (GBLOC MAPS).
- **Zone IV** shipments will be processed by USCG Base Ketchikan (GBLOC MAPK) with the exception of Cordova which will be processed by USCG Base Kodiak (GBLOC MAPS).
- **Zone V**  -- NSRA15 -- search will not work -- reference file for known NSRA15 locations:  
[NSRA15-duty-locations-data-1731515178779.csv](https://github.com/user-attachments/files/17739099/NSRA15-duty-locations-data-1731515178779.csv)

**Zone 2 affiliated duty locations are the ones to pay attention to!** Use TRDM file as reference to determine Zone affiliation.
[(CUI) TRDM AK Duty Locations.xlsx](https://github.com/user-attachments/files/17738888/CUI.TRDM.AK.Duty.Locations.xlsx)

3. To make testing a little easier first login using Air Force or Space Force affiliated service member. Sort the TRDM file for all ZONE 2. Do spot checks  and verify the duty locations selected display the correct counseling offices
`select name from transportation_offices where gbloc = 'MBFL';`

I've added DEBUG logging to help testing

**Successful Duty Location Request:**
2024-11-13T20:43:12.062Z	DEBUG	transportation_office/transportation_office_fetcher.go:285	Found specific department match -- serviceMember.Affiliation: AIR_FORCE, DutyLocaton: Fairbanks, AK 99710, GBLOC: MBFL, departmentIndicator: AIR_AND_SPACE_FORCE, RateAreaName: Alaska (Zone) II, dutyLocation.Address.ID: 33f9674d-0243-4c7e-9759-ca7f2c52f46a	{"git_branch": "INT-B-21608-Alaska-Update-search-counseling-office-query", "git_commit": "f74a7071ad25b499d7cd6568ac804f95fd603a7d", "host": "milmovelocal:3000", "milmove_trace_id": "8307b5be-ebce-4eb9-ba6b-a8645dc0e981", "session_id": "CXpIw9j3BnGlEVbeDbUagOeObAUN57pAxc-TxWmZM40"}

**Pay attention to GBLOC, departmentIndicator, RateAreaName information in the log output. It should match the business logic requirements for the Zone.**

**unsuccessful/NSRA15 Duty Location Request:**
2024-11-13T20:46:05.985Z	ERROR	internalapi/transportation_offices.go:79	Error searching for Counseling Offices: 	{"git_branch": "INT-B-21608-Alaska-Update-search-counseling-office-query", "git_commit": "f74a7071ad25b499d7cd6568ac804f95fd603a7d", "host": "milmovelocal:3000", "milmove_trace_id": "7ed5738e-d8d5-4516-8ccb-a3884e6e3c2d", "session_id": "CXpIw9j3BnGlEVbeDbUagOeObAUN57pAxc-TxWmZM40", "error": "ImplementationError: Error: Cannot determine GBLOC -- serviceMember.Affiliation: AIR_FORCE, DutyLocaton: Adak, AK 99546, departmentIndicator: AIR_AND_SPACE_FORCE, dutyLocation.Address.ID: 78051891-ee44-4a61-9dab-4c1fd0344a76"}

4. Login using ARMY affiliated service member. Follow steps outlined in step 3.
5. FF for Hawaii was added to this BL. Search for known HI locations and verify there are no matches.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### CODE REVIEW

File: pkg/models/re_oconus_rate_areas.go [hunk 1]
```diff
@@ -7,13 +7,13 @@ import (
...
 
 type OconusRateArea struct {
-	ID                 uuid.UUID `json:"id" db:"id" rw:"r"`
-	RateAreaId         uuid.UUID `json:"rate_area_id" db:"rate_area_id" rw:"r"`
-	CountryId          uuid.UUID `json:"country_id" db:"country_id" rw:"r"`
-	UsPostRegionCityId uuid.UUID `json:"us_post_region_city_id" db:"us_post_region_city_id" rw:"r"`
-	Active             bool      `json:"active" db:"active" rw:"r"`
-	CreatedAt          time.Time `json:"created_at" db:"created_at" rw:"r"`
-	UpdatedAt          time.Time `json:"updated_at" db:"updated_at" rw:"r"`
+	ID                 uuid.UUID `json:"id" db:"id"`
```
Comment (r-mettler) [inline pkg/models/re_oconus_rate_areas.go:10]: add `rw:"r"` to each of these fields, we don't want the app to even attempt to update data in this table

Comment (MInthavongsay) [reply pkg/models/re_oconus_rate_areas.go:10]: I had to remove the readonly flag to have the ability to insert records to create test data. Is there another way?

**Comment by r-mettler** [comment]
Spot checks with NSRA15 Duty Locations displayed error message in console as expected
Spot checks of non Zone 5 Duty Locations looked good

File: pkg/models/re_oconus_rate_areas.go (hunk 1 above)
Comment (r-mettler) [reply pkg/models/re_oconus_rate_areas.go:10]: For tables that we have static data we want to add read only and don't need to test the insert of data because the app should never insert data into these tables. If you however find that you need to access data in the tables then make sure the table is listed in the `scripts/db-truncate` file to tell the tests to not remove the data from the table. `re_oconus_rate_areas` should already be in the truncate file.

Comment (r-mettler) [reply pkg/models/re_oconus_rate_areas.go:10]: Since this is high priority I am fine with keeping this like it is for now to get this PR pushed through and fixing this in the next BL scheduled to fix the Counseling Office dropdown bug mentioned in the PR description.

**Review by r-mettler — APPROVED** [review]

File: .envrc [hunk 2]
```diff
@@ -157,7 +157,10 @@ export FEATURE_FLAG_MOBILE_HOME=true
 export FEATURE_FLAG_UNACCOMPANIED_BAGGAGE=false
 
 # Feature flag to allow AK to be entered as a state
-export FEATURE_FLAG_ENABLE_ALASKA=false
+export FEATURE_FLAG_ENABLE_ALASKA=true
```
[the remark this answers is not in the document]
Comment (MInthavongsay) [reply .envrc:160]: It's only true locally. When this goes to INT and Dan Jackson is ready to start testing is when we flip the FF to true.

File: pkg/services/transportation_office/transportation_office_fetcher.go [hunk 3]
```diff
@@ -171,3 +233,71 @@ func findCounselingOffice(appCtx appcontext.AppContext, dutyLocationID uuid.UUID
...
+
+	query := appCtx.DB().Q().RawQuery(sqlQuery, dutyLocation.Address.ID)
+	err = query.All(&oconusGblocDepartmentIndicator)
+	if err != nil {
+		return nil, err
+	}
+
+	// Determine departmentIndicator based on service member's affiliation
+	var departmentIndicator *string = nil
+	if serviceMember.Affiliation != nil && (*serviceMember.Affiliation == models.AffiliationAIRFORCE || *serviceMember.Affiliation == models.AffiliationSPACEFORCE) {
```
[the remark this answers is not in the document]
Comment (antgmann) [reply pkg/services/transportation_office/transportation_office_fetcher.go:268]: @traskowskycaci After checking some of my notes, maybe I'm jumping the gun with saying it's planned work. I've briefly discussed the separate enums and how they work with @begrohmann - she has expressed the possibility that this may need to occur in order to validate gblocs. 

It may not actually be necessary to alter this in the future if:
 1. Orders affiliations always align with branches of service and gblocs in such a manner that the above conditionals always evaluate correctly, and 
 2. We don't go forward with aligning those two data types. 
 
However if such a feature were to occur in the future, these conditionals should be checked to make sure everything aligns nicely.

File: .envrc (hunk 2 above)
[the remark this answers is not in the document]
Comment (antgmann) [reply .envrc:160]: You're right! I thought this was a config env file. My bad

File: pkg/services/transportation_office/transportation_office_fetcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (MInthavongsay) [reply pkg/services/transportation_office/transportation_office_fetcher.go:268]: Anthony -- are you talking about AIR_AND_SPACE_FORCE not lining up to AIR_FORCE or SPACE_FORCE?  I coded the BL according to what the data looks now according to Beth's migration work. Right now, we only group AIR_FORCE and SPACE_FORCE together using AIR_AND_SPACE_FORCE as departmentIndicator. My logic mirrors what is defined in the  DepartmentIndicator enum:
Screenshot 2024-11-14 at 2 01 16 PM

**Review by antgmann — APPROVED** [review]
Only issue I'm running into has to do with the caching of counseling offices upon selection of a valid location such that they're still available when selecting an invalid location. However, we've discussed this in a huddle and it should be resolved with the completion of the downstream mentioned in the PR description. I haven't managed to break it in any other way, and all the locations seem to align correctly.

File: pkg/services/transportation_office/transportation_office_fetcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (traskowskycaci) [reply pkg/services/transportation_office/transportation_office_fetcher.go:268]: @antgmann I think it's worth discussing how we would verify your #1 there of "Orders affiliations always align with branches of service and gblocs in such a manner that the above conditionals always evaluate correctly"

I agree that's a good point - but it is probably safer to explicitly enforce that somehow (IF we need to keep both distinct enums)

Just thinking out loud here - that's an area I could see giving us trouble if we have two separate enums, meaning the same thing, but not quite 100% the same underlying values

[the remark this answers is not in the document]
Comment (traskowskycaci) [reply pkg/services/transportation_office/transportation_office_fetcher.go:268]: Not for this BL, that is. Just thinking ahead for the future

File: .envrc (hunk 2 above)
[the remark this answers is not in the document]
Comment (traskowskycaci) [reply .envrc:160]: I do think this should be off until we are ready to turn Alaska on overall - see above where we are keeping the UB flag off as well

File: pkg/services/transportation_office/transportation_office_fetcher.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (antgmann) [reply pkg/services/transportation_office/transportation_office_fetcher.go:268]: @traskowskycaci I think since the gbloc locations are created with DepartmentIndicator, and only with department indicators that include two potential service member affiliations that line up nicely with the way @MInthavongsay assigned them, that #1 is good, and this pr is good. Issues with this would only appear if a duty location was given a department indicator that cannot directly have a service member affiliation mapped to it, such as office of secretary of defense, which doesn't line up nicely between the two.

tldr; I concur with your thoughts and don't really like how these two enums mean the same thing with different values and get assigned based on one another. It gets confusing when you lump them together, and when you keep them separate. Can't win I guess 🤷 

@MInthavongsay Your implementation doesn't have anything wrong with it. What I'm talking about is that Department Indicator and Service Member Affiliation being assigned based on another can get tricky because of the nature of the two enums, and in the future they may be combined. I thought this was happening for sure. But this may or may not happen. If it does, I'm just trying to document that it'll have to change these conditionals. Nothing to worry about for now 👍
